### ISSUE TITLE
Unresolved references with third party libstd

### ISSUE THREAD

**Issue by elichai** [intellij-rust/intellij-rust#2561]
Hi,
If I use a third party `std` library the plugin doesn't recognize any functions.
e.g.
```rust
extern crate sgx_tstd as std;
use std::string::String;
use std::vec::Vec;

let mut vec = Vec::new();
let hello = String::from("Hello, world!");
let mut hello = String::with_capacity(25);
```
Here both `Vec` and `String` are red and shows `Unresolved Reference`
And if I replace `extern crate sgx_tstd as std` with regular `extern crate std` everything works.

Thanks,
Elichai.

**Comment by JakubAdamWieczorek** [comment]
I don't think there is a bug here, as the code as it is won't compile anyway (here's a simplified version).

```rust
extern crate sgx_tstd as std;

use std::string::String;

pub fn f() {
    let s = String::from("string");
}
```

```
error[E0259]: the name `std` is defined multiple times
  |
  = note: `std` must be defined only once in the type namespace of this module

error: duplicate lang item in crate `sgx_tstd`: `f32`.
  |
  = note: first defined in crate `std`.

error: duplicate lang item in crate `sgx_tstd`: `f64`.
  |
  = note: first defined in crate `std`.

error: duplicate lang item in crate `sgx_tstd`: `panic_fmt`.
  |
  = note: first defined in crate `std`.
```

If you want to use a third-party standard library, you need a `#![no_std]` attribute on the crate to disable the implicit use of the default standard library. With it in place IntelliJ Rust resolves the symbols properly:

```rust
#![no_std]

extern crate sgx_tstd as std;

use std::string::String;

pub fn f() {
    let s = String::from("string");
}
```

The funny thing is that this **will** build with `cargo check` on the command line but still **won't** build n IntelliJ as that uses `cargo check --all --all-targets`, and both the `tests` and `examples` targets won't build in the `no_std` regime. But that's a separate issue… The `no_std` configuration has had many of those anyway.

**Comment by elichai** [comment]
Sorry I forgot to mention this, but of course I have `!#[no_std]` in my lib.rs

**Comment by elichai** [comment]
Oh I see what the problem is.
I had `#![cfg_attr(not(target_env = "sgx"), no_std)]` and jetbrains doesn't use it in checking so from it's prespective there is no `#![no_std]`.

**Comment by JakubAdamWieczorek** [comment]
@elichai, okay, that makes sense! 🙂 The plugin is target-agnostic when it does type checking and name resolution, but in principle those should never give false negatives (i.e. indicate the code won't compile when in reality it does). False positives are acceptable. So yes, this is a bug. This is a minimal example:

**src/lib.rs**
```rust
#![cfg_attr(not(target_os="netbsd"), no_std)]
extern crate rand as std;
pub use std::random;
pub fn f() {
    let _: f64 = random(); // `random` does not resolve here, but it should.
}
```

**Cargo.toml**
```toml
[package]
name = "test_nostd"
version = "0.1.0"

[dependencies]
rand = "*"
```

If we change the alias from `std` to whatever else, `random` resolves properly.

**Comment by JakubAdamWieczorek** [comment]
I think it's more straightforward than that. This issue isn't really specific to `std`. It occurs any time `extern crate` occurs with the same name (or alias):

```rust
#[cfg(target_os = "linux")]
extern crate rand;
#[cfg(target_os = "windows")]
extern crate rand;

use rand::random;

pub fn f() {
    let _v: u64 = random(); // does not resolve.
}
```

Of course, this won't compile but with the plugin's limitations around processing `#[cfg_attr]` it seems it should be more tolerant here. As it is, this test case results in a false negative.

**Comment by elichai** [comment]
Hi, Wanted to know if you have ideas on how to solve this.
Thanks!

**Comment by ortem** [comment]
This issue was fixed by #4537
